src/atomEffect2.ts

```diff
@@ -0,0 +1,68 @@
+import type { Getter, Setter, WritableAtom } from 'jotai/vanilla'
+import { atom } from 'jotai/vanilla'
+
+type CleanupFn = () => PromiseOrValue<void>
+type PromiseOrValue<T> = Promise<T> | T
+
+export function atomEffect(
+  effectFn: (get: Getter, set: Setter) => PromiseOrValue<void | CleanupFn>
+) {
+  const refAtom = atom<{
+    mounted: boolean
+    rerun: boolean
+    inProgress: number
+    cleanup: CleanupFn | void
+  }>({
+    mounted: false,
+    rerun: false,
+    inProgress: 0,
+    cleanup: undefined,
+  })
+  const initAtom = atom(null, (get, _set, mounted: boolean) => {
+    const ref = get(refAtom)
+    if (mounted) {
+      ref.mounted = true
+    } else {
+      ref.cleanup?.() // do not await
+      ref.mounted = false
+      ref.rerun = false
+      ref.cleanup = undefined
+    }
+  })
+  initAtom.onMount = (init) => {
+    init(true)
+    return () => init(false)
+  }
+
+  const effectAtom = atom(
+    async (get, { setSelf }) => {
+      const ref = get(refAtom)
+      if (!ref.mounted) {
+        return
+      }
+      if (ref.inProgress) {
+        ref.rerun = true
```

Comment:
inProgress indicates that the effectFn causes a write. If we reach this line, then we have an infinite loop scenario.

It means the write caused the atomEffect read to re-fire. Even though we return here, since rerun is now true, the while loop would rerun. If the condition that triggered the set call was not based on the value that was changed by the set call, this would be an infinite loop.